### PR TITLE
!!! FEATURE: Introduce flushByTags to taggable cache backends

### PR DESCRIPTION
**Pull request by Sebobo**
**What I did**

This change introduces the `flushByTags` method to taggable cache backends. This allows each backend to be further optimised in the future to find the best way to flush multiple tags from its storage.

Resolves: #2717

**How I did it**

The `TaggableBackendInterface` and `FrontendInterface` were extended with the method `flushByTags` and each provided backend in the core got a simple implementation for it to fulfil the interface.

### CODE REVIEW

**Comment by Sebobo** [comment]
Still have to update the tests

File: Neos.Cache/Classes/Frontend/AbstractFrontend.php [hunk 1]
```diff
@@ -152,6 +152,27 @@ public function flushByTag(string $tag): int
...
+     * @throws \InvalidArgumentException
+     * @api
+     */
+    public function flushByTags(array $tags): int
+    {
+        foreach ($tags as $tag) {
+            if (!$this->isValidTag($tag)) {
+                throw new \InvalidArgumentException('"' . $tag . '" is not a valid tag for a cache entry.', 1646209443);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (Sebobo) [reply Neos.Cache/Classes/Frontend/AbstractFrontend.php:169]: I actually also wondered why we need this. If the tag is invalid it wouldn't hurt anyway.

Maybe there is a use case to prevent SQL injections or something?
But then I agree too, that the backend should take care of that.

[the remark this answers is not in the document]
Comment (bwaidelich) [reply Neos.Cache/Classes/Frontend/AbstractFrontend.php:169]: It might even be that a tag that is valid according to the frontend is dangerous in the backend (it should not, but it _might_) so IMO this _has_ to be tested/sanitized in the backend anyways

[the remark this answers is not in the document]
Comment (Sebobo) [reply Neos.Cache/Classes/Frontend/AbstractFrontend.php:169]: Agree. But should I really do this in this PR too?

**Comment by mficzel** [comment]
You can reduce the breakiness to almost zero by implementing a primitive flushByTags method in the abstractBackend.

File: Neos.Cache/Classes/Frontend/AbstractFrontend.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (mficzel) [reply Neos.Cache/Classes/Frontend/AbstractFrontend.php:169]: There is no need to be smarter than the current solution. As this is done already in flushByTag it does not produce extra costs we do not already pay.

**Review by mficzel — APPROVED** [review]
Fine by reading. Pretty safe aswell as no functionality is changed yet.

File: Neos.Flow/Classes/Mvc/Routing/RouterCachingService.php [hunk 2]
```diff
@@ -206,6 +206,17 @@ public function flushCachesByTag($tag)
         $this->resolveCache->flushByTag($tag);
     }
 
+    /**
+     * Flushes 'findMatchResults' and 'resolve' caches for the given $tags
+     *
+     * @param array<string> $tags
+     */
+    public function flushCachesByTags(array $tags): void
```
Comment (bwaidelich) [inline Neos.Flow/Classes/Mvc/Routing/RouterCachingService.php:214]: if we introduce this method, we should IMO use it in the core and deprecate (or even remove) `flushCachesByTag()` but I would suggest to do this in a separate PR

File: Neos.Cache/Classes/Frontend/AbstractFrontend.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (bwaidelich) [reply Neos.Cache/Classes/Frontend/AbstractFrontend.php:169]: > There is no need to be smarter than the current solution

I beg to differ: We introduce this method exclusively for performance reasons, so IMO we can decide to be smarter than before :)
But I was expecting our backends to validate the tag and not all of them do.. So this would be too much of a change for one PR I guess

**Review by bwaidelich — COMMENTED** [review]
+1 by reading, but 3 nitpicks/questions:
1. somewhere @mficzel suggested to introduce the trivial implementation of `flushByTags()` in the `AbstractBackend` for a greater backwards compatibility. Was that idea discarded intentionally?
2. I'd suggest not to change the `RouterCachingService` with this PR (see inline comment)
3. (nitpick): I'd suggest to remove the checklist from the PR description so that it doesn't end up in the merge commit (and thus in the changelog)

File: Neos.Flow/Classes/Mvc/Routing/RouterCachingService.php (hunk 2 above)
Comment (Sebobo) [reply Neos.Flow/Classes/Mvc/Routing/RouterCachingService.php:214]: It is only used in Neos.Neos and contained in the PR https://github.com/neos/neos-development-collection/pull/3631

I don't see the necessity to remove the option to flush a single tag.

File: Neos.Cache/Classes/Frontend/AbstractFrontend.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (Sebobo) [reply Neos.Cache/Classes/Frontend/AbstractFrontend.php:169]: I can prepare a 2nd PR for moving the validation to the backends.

**Comment by mficzel** [comment]
@bwaidelich i scrapped the idea of a trivial flushByTags implementation myself because there is no base class for taggable Backends that could implement this. So this would require quite a bit of restructuring that is not required for this pr.

**Comment by bwaidelich** [comment]
> i scrapped the idea of a trivial flushByTags implementation myself because there is no base class for taggable Backends that could implement this

I thought something like

```php
class AbstractBackend {
   // ...
    public function flushByTags(array $tags): int
    {
        if (!$this instanceof TaggableBackendInterface) {
           // throw or return 0
        }
        $flushed = 0;
        foreach ($tags as $tag) {
            $flushed += $this->flushByTag($tag);
        }
        return $flushed;
    }
```

but.. Not sure if the B/C is worth the weird construct in this case

File: Neos.Cache/Classes/Backend/TaggableBackendInterface.php [hunk 3]
```diff
@@ -29,6 +29,15 @@ interface TaggableBackendInterface extends BackendInterface
...
     public function flushByTag(string $tag): int;
 
+    /**
+     * Removes all cache entries of this cache which are tagged by any of the specified tags.
+     *
+     * @param array<string> $tags The tags the entries must have
+     * @return integer The number of entries which have been affected by this flush or NULL if the number is unknown
+     * @api
+     */
+    public function flushByTags(array $tags): int;
```
[the remark this answers is not in the document]
Comment (Sebobo) [reply Neos.Cache/Classes/Backend/TaggableBackendInterface.php:39]: thx, good point, then this is incorrect since 2017 when the type hint for `flushByTag` was added. I copied the description without noticing the mention of the NULL value. Therefore I would adjust the docs for both methods.

**Comment by Sebobo** [comment]
> > i scrapped the idea of a trivial flushByTags implementation myself because there is no base class for taggable Backends that could implement this
> 
> I thought something like
> 
> ```
> class AbstractBackend {
>    // ...
>     public function flushByTags(array $tags): int
>     {
>         if (!$this instanceof TaggableBackendInterface) {
>            // throw or return 0
>         }
>         $flushed = 0;
>         foreach ($tags as $tag) {
>             $flushed += $this->flushByTag($tag);
>         }
>         return $flushed;
>     }
> ```
> 
> but.. Not sure if the B/C is worth the weird construct in this case

I don't feel good about adding tagging related functionality to the AbstractBackend. I think the handful of custom backend implementations can adjust their code without much effort. We anyway have a super simple "sample implementation" in the existing backends and can add it to the upgrade guide. And packages like Sandstorms optimised Redis backend can already be updated without the need to introduce a major version.

**Review by kdambekalns — APPROVED** [review]

**Comment by Sebobo** [comment]
Thx everyone!
